### PR TITLE
feat(subtitles): store downloaded subtitles in either storage backend

### PR DESCRIPTION
**Pull request by Quick104**
## Problem

Related issue: #774 — extends the S3-optional storage work past artwork.

Subtitles were the only client-visible feature a bucket-less install lost outright. The manager was
built solely when a public S3 bucket existed (`internal/api/router.go`), so without one there was no
subtitle search, download, upload, AI publication, or playback of a downloaded track — on the native
API or through jellycompat.

Stacked on #1112. Review that one first.

## Approach

Replace the `subtitles.S3Client` interface with a key-only `BlobStore` that either backend satisfies
through `blobstore.NewByteStore`, and gate the manager on having somewhere to write rather than on S3
specifically.

The interface is key-only — no bucket, no signed URLs — because subtitle bytes always flow through
the server. `GetSubtitleContent` reads the object and writes the body; a client never receives a
signed storage URL for a subtitle. That makes this a storage swap with no API change and nothing for
`silo-apple` or `silo-android` to do. I checked both: neither references the subtitle storage
surface.

Subtitles keep the `subtitles/` key prefix, which is what separates them from artwork when a local
backend puts both in one root.

`ByteStore.Delete` checks the deleted count rather than discarding it. S3 batch deletion reports
per-object failures in its response body and still returns a nil error with a short count, so a
failed delete read as success — and because subtitle deletion removes the row first, the manager's
cleanup warning never fired and the object was orphaned. Review caught this. An absent key still
counts as deleted, so cleanup after a failed publish stays quiet.

The `downloaded_subtitles.s3_key` column and the `S3Key` Go field keep their names. They hold a
backend-neutral object key now, and renaming them would touch every query in
`internal/subtitles/pgrepo.go` for no functional gain.

## Validation

- `go build ./...`, `go vet ./...`, `gofmt -l .` clean.
- `go test` for `internal/subtitles/...`, `internal/api/handlers/...`, `internal/jellycompat/...`,
  `internal/apiv2/...`.
- New tests: a publish, re-read, and cleanup round trip through a real filesystem store in a temp
  directory, asserting the object lands under `subtitles/<media file id>/` and is on disk; and that
  deleting an absent key is not an error.
- End to end against a running server with no S3 configured: the subtitle provider list and the
  admin subtitle surface both answer, where previously the manager would not have been constructed.

One test fake caught a real problem worth mentioning. `viewerDeleteObjects` in
`subtitle_delete_test.go` embeds the shared fake and overrode `DeleteObject`. After the rename it
still compiled — it just stopped overriding anything, so the deletion counter never incremented and
`TestViewerSubtitleDeletionAuthorityAndCAS` failed. Renaming a method on an embedded interface is
silent at compile time; the test caught it, which is the argument for keeping that assertion.

Pre-existing and unrelated: `TestAdminResourceCapabilitiesAndScope` in `internal/apiv2` fails on
macOS because the development machine has no cgroups. Confirmed identical at the merge base.

## Risks

No API change, no schema change, no migration, no client work.

An S3 deployment keeps its bucket and every existing key; the manager is constructed from the same
public bucket it used before, reached through the assets store.

## Checklist

- [x] I read and can explain the complete diff.
- [x] This pull request addresses one concern.

## AI Disclosure

- Harness: Claude Code (T3 Code)
- Tool(s): Claude Code; Claude Code subagent used for adversarial plan review
- Model(s): `claude-opus-5[1m]` (Opus 5, 1M context) for implementation; Fable for the review pass
- Involvement: Fully AI-generated, human verified
- Adversarial review: A separate model reviewed the implementation plan against the source before
  any code was written. Its findings were concentrated in the first and third PRs of this stack and
  are documented there; it raised nothing specific to the subtitle change beyond confirming that
  subtitle delivery never presigns, which is what makes this swap invisible to clients. That claim
  was verified against the read paths in `internal/api/handlers/stream.go` and
  `internal/jellycompat/streams.go`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

> [!NOTE]
> ### Replace S3 client with generic `BlobStore` for downloaded subtitles
> - Introduces `subtitles.BlobStore` (key-only `Put`, `Get`, `Delete`) to replace direct S3 client and bucket dependencies across the subtitle manager, API stream handlers, and Jellyfin compatibility layer.
> - Adds `blobstore.ByteStore` adapter in [bytes.go](https://github.com/Silo-Server/silo-server/pull/1113/files#diff-36ba7a949d70aae8a28c39346005dbfdd8beefa8c9dcc2caa440eaf89378bb9a) that wraps the configured assets `Store` so subtitle bytes can be stored on either filesystem or S3 backends.
> - Wires `deps.Blobs.Assets` into the subtitle manager and stream handlers via `ByteStore`, removing the hard requirement on `S3Public` for subtitle serving.
> - Updates setup wizard and infrastructure settings UI to mention that subtitles are stored with library assets.
> - Risk: `ByteStore.Delete` in [bytes.go](https://github.com/Silo-Server/silo-server/pull/1113/files#diff-36ba7a949d70aae8a28c39346005dbfdd8beefa8c9dcc2caa440eaf89378bb9a) requires the underlying batch-delete to report exactly one deleted object, returning an error even if the underlying call succeeded but the count differs.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 1e0d49d.</sup>
> 
>

### CODE REVIEW

File: internal/blobstore/bytes.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+func (b *ByteStore) Put(ctx context.Context, key string, data []byte) error {
+	return b.store.Put(ctx, key, data)
+}
+
+func (b *ByteStore) Get(ctx context.Context, key string) ([]byte, error) {
+	return GetBytes(ctx, b.store, key)
+}
+
+func (b *ByteStore) Delete(ctx context.Context, key string) error {
+	_, err := b.store.Delete(ctx, []string{key})
```
[the remark this answers is not in the document]
Comment (Quick104) [reply internal/blobstore/bytes.go]: Confirmed and fixed in 1e0d49dedf3a.

Verified against `s3client.DeleteObjects`: it returns a nil error on both paths that can lose an object. The batch path subtracts `len(out.Errors)` from the count and only logs, and the per-key fallback `continue`s on failure without counting. Discarding the count therefore reported success while the object remained, and because subtitle deletion removes the row first, `cleanupSubtitleObject`'s warning never fired.

`ByteStore.Delete` now returns an error when the count is not one. An absent key still counts as deleted, matching the store contract, so cleanup after a failed publish stays quiet. Added `TestByteStoreDeleteReportsAShortCountAsFailure` with a store that returns `(0, nil)`, plus a case pinning the absent-key behavior.

File: internal/api/router.go [hunk 2]
```diff
@@ -1284,11 +1284,12 @@ func newChiRouter(deps Dependencies) chi.Router {
 		}
 	}
 
-	// Wire subtitle repo and S3 client onto streamHandler for S3-stored subtitle serving.
-	if streamHandler != nil && subtitleRepo != nil && deps.S3Public != nil {
+	// Wire the subtitle repo and blob store onto streamHandler so downloaded
+	// subtitles serve from whichever backend stores them.
+	subtitleBlobs := blobstore.NewByteStore(deps.Blobs.Assets)
```
Comment (Quick104) [inline internal/api/router.go:1324]: **Medium, needs author decision: `local` backend plus a public S3 bucket moves subtitles away from their existing keys.**

Before this PR the subtitle manager always used `deps.S3Public`. Now it uses `deps.Blobs.Assets`, and `blobstore.Open` builds a filesystem store whenever `artwork.storage_backend` is `local`, even when a public bucket is configured. The recorded-identity check only guards the assets store, so nothing stops this switch.

Trigger: an install with `artwork.storage_backend=local` and `s3.public_bucket` set, with downloaded subtitles already in the bucket. After upgrading:

- Every existing `downloaded_subtitles` row points at a key that is only in S3. Native serving returns 502 `s3_error`, and jellycompat returns 500.
- `StoreSubtitle` legacy-duplicate verification fails with `verify legacy subtitle`, so downloading the same content again errors.
- Deleting an old subtitle removes the row. The local delete counts the absent key as deleted, which leaves the S3 object orphaned without the reconciliation warning.

This contradicts the Risks section ("An S3 deployment keeps its bucket and every existing key"). The exposure is small: the local artwork backend (d8fb5ca89) has only been on `main` since 2026-09-15 and no release has shipped it. Still, dev installs in that configuration will hit it.

Options:
1. Accept the move, since it's pre-release, and correct the Risks section to say `local` plus S3 installs lose access to earlier subtitles. That matches the new settings copy ("Where Silo keeps artwork, subtitles, …").
2. Keep subtitles on `S3Public` whenever it is configured and fall back to `Assets` only when there is no bucket. This preserves every existing key, but then the backend setting no longer controls where subtitles go.

I lean toward option 1 given there's no release, but it's your call. I haven't changed this.

**Review by Quick104 — COMMENTED** [review]
Review of `1e0d49d`. One design concern needs an author decision (inline). One small copy fix, which I'm handling in a follow-up commit on this branch:

**Low, fixing: stale admin empty state.** `web/src/components/admin/subtitles/AdminSubtitlesTable.tsx:85` still tells admins that uploads and downloads "will appear here once subtitles are stored in S3." After this PR, a bucket-less install stores subtitles on local disk, so the copy suggests S3 is required when it isn't. I'm dropping the S3 clause.

Checked and fine: nil `*ByteStore` never reaches an interface field (every assignment is guarded); the filesystem sweeper only removes `.tmp-`/`.probe-` files and image-cache cleanup only deletes artwork prefixes, so subtitle objects in a shared local root are safe; legacy `subtitles/<id>/<lang>_<provider>_<hash>.<fmt>` keys pass `ValidateKey`; absent-key deletes count as 1 on both backends; subtitle writes through the recording wrapper lock the backend identity like artwork writes do.

File: internal/api/router.go (hunk 2 above)
Comment (Quick104) [reply internal/api/router.go:1324]: Deferred by maintainer decision. Migration between S3 and local storage is handled in a separate PR that waits on #1113 and #1114, so this PR doesn't need to preserve subtitle keys when the backend changes.

A note for that migration: installs that were already on `local` with a public bucket configured won't change backends, but their earlier subtitles are still in S3. The migration needs to copy `subtitles/` objects out of the public bucket for those installs too, not only when the backend setting changes. Resolving this thread here.

**Comment by Quick104** [comment]
Review summary for head `6879f4e`.

**Addressed**
- Stale admin empty state: fixed in 6879f4e15. `AdminSubtitlesTable` no longer tells admins subtitles appear "once subtitles are stored in S3". Prettier is clean and the admin web tests pass (68 files, 690 tests).
- Keeping subtitle keys readable on a `local` backend with a public bucket configured: deferred by maintainer decision to the separate S3/local migration PR. The thread records the edge case that migration needs to cover.
- The earlier Codex finding on `ByteStore.Delete`: I checked the fix in 1e0d49d against `s3client.DeleteObjects`. A short count now returns an error, absent keys still count as deleted, and both cases have tests.

**Validation:** `go build ./...` and `go vet` are clean. Go tests pass for `internal/blobstore`, `internal/subtitles`, `internal/api/handlers`, and `internal/jellycompat`. CI on `6879f4e` is pending.

**Remaining:** #1112 has to merge first. The Risks line "An S3 deployment keeps its bucket and every existing key" doesn't hold for `local` plus S3 installs until the migration lands; consider noting that in the description.

**Mergeability: 8/10 — conditional.** The storage swap is clean and gives bucket-less installs subtitles without API, schema, or client changes. Conditions: green CI on `6879f4e` and #1112 merging first.

File: internal/api/router.go [hunk 3]
```diff
@@ -1319,11 +1319,12 @@ func newChiRouter(deps Dependencies) chi.Router {
...
```
[the remark this answers is not in the document]
Comment (Quick104) [reply internal/api/router.go:1324]: Confirmed against `e72e19c9`. `blobstore.Open` builds a filesystem assets store whenever the backend is `local`, even with a public bucket configured, and the router passes `deps.Blobs.Assets` to the subtitle manager, the stream handler, and jellycompat. Rows written to S3 before the upgrade are unreachable on those installs.

This was raised and settled earlier in [this thread](https://github.com/Silo-Server/silo-server/pull/1113#discussion_r4074252350). By maintainer decision, moving data between S3 and local storage belongs to the separate migration PR, which is waiting on #1113 and #1114. That PR needs to cover this case: installs already on `local` with a public bucket keep their older subtitles in S3 and never change backends, so the migration has to copy `subtitles/` objects out of the bucket for them too. Nothing released ships the local backend (it landed on `main` on 2026-09-15), so no published install is affected in the meantime. Resolving here.

File: internal/blobstore/bytes.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Quick104) [reply internal/blobstore/bytes.go]: Verified at `e72e19c9`: `ByteStore.Delete` returns an error unless the store reports exactly one object deleted (`internal/blobstore/bytes.go`), and `TestByteStoreDeleteReportsAShortCountAsFailure` and `TestByteStoreDeleteAcceptsAnAbsentKey` cover the short-count and absent-key cases. Resolving.
